src/qibo/models/dbi/utils.py

```diff
@@ -0,0 +1,165 @@
+    mode: DoubleBracketGeneratorType = DoubleBracketGeneratorType.single_commutator,
+):
+    """Selects the best double bracket rotation generator from a list and runs the
+
+    Args:
+        dbi_object (_DoubleBracketIteration): The target DoubleBracketIteration object.
+        d_list (list): List of diagonal operators (np.array) to run from.
+        step (float): Fixed iteration duration.
+            Defaults to ``None``, uses hyperopt.
+        step_min (float): Minimally allowed iteration duration.
+        step_max (float): Maximally allowed iteration duration.
+        max_evals (int): Maximally allowed number of evaluation in hyperopt.
+        compare_canonical (bool): If `True`, the optimal diagonal operator chosen from "d_list" is compared with the canonical bracket.
+        mode (_DoubleBracketGeneratorType): DBI generator type used for the selection.
+
+    Returns:
+        The updated dbi_object, index of the optimal diagonal operator, respective step duration, and evolution direction.
+    """
+    norms_off_diagonal_restriction = [
+        dbi_object.off_diagonal_norm for _ in range(len(d_list))
+    ]
+    optimal_steps = [0 for _ in range(len(d_list))]
+    flip_list = [1 for _ in range(len(d_list))]
+    for i, d in enumerate(d_list):
+        # prescribed step durations
+        dbi_eval = deepcopy(dbi_object)
+        flip_list[i] = CS_angle_sgn(dbi_eval, d)
+        if flip_list[i] != 0:
+            if step is None:
+                step_best = dbi_eval.hyperopt_step(
+                    d=flip_list[i] * d,
+                    step_min=step_min,
+                    step_max=step_max,
+                    space=hp.uniform,
+                    optimizer=tpe,
+                    max_evals=max_evals,
+                )
+            else:
+                step_best = step
+            dbi_eval(step=step_best, d=flip_list[i] * d)
+            optimal_steps[i] = step_best
+            norms_off_diagonal_restriction[i] = dbi_eval.off_diagonal_norm
+    # canonical
+    if compare_canonical is True:
+        flip_list.append(1)
+        dbi_eval = deepcopy(dbi_object)
+        dbi_eval.mode = DoubleBracketGeneratorType.canonical
+        if step is None:
+            step_best = dbi_eval.hyperopt_step(
+                step_min=step_min,
+                step_max=step_max,
+                space=hp.uniform,
+                optimizer=tpe,
+                max_evals=max_evals,
+            )
+        else:
+            step_best = step
+        dbi_eval(step=step_best)
+        optimal_steps.append(step_best)
+        norms_off_diagonal_restriction.append(dbi_eval.off_diagonal_norm)
+    # find best d
+    idx_max_loss = norms_off_diagonal_restriction.index(
+        min(norms_off_diagonal_restriction)
+    )
+    flip = flip_list[idx_max_loss]
+    step_optimal = optimal_steps[idx_max_loss]
+    dbi_eval = deepcopy(dbi_object)
+    if idx_max_loss == len(d_list) and compare_canonical is True:
```

Comment:
Here I guess that you could just check if `d` is $\Delta(H)$, right? 
It should improve readability